source/common/upstream/load_balancer_config.h

```diff
@@ -0,0 +1,281 @@
+};
+
+/**
+ * Load balancer config that used to wrap typed maglev config.
+ */
+class TypedMaglevLbConfig : public Upstream::LoadBalancerConfig {
+public:
+  TypedMaglevLbConfig(const envoy::extensions::load_balancing_policies::maglev::v3::Maglev& config);
+
+  const envoy::extensions::load_balancing_policies::maglev::v3::Maglev lb_config_;
+};
+
+/**
+ * Load balancer config that used to wrap legacy ring hash config.
+ */
+class LegacyTypedRingHashLbConfig : public Upstream::LoadBalancerConfig {
+public:
+  LegacyTypedRingHashLbConfig(const envoy::config::cluster::v3::Cluster& config);
+
+  absl::optional<envoy::config::cluster::v3::Cluster::RingHashLbConfig> lb_config_;
+};
+
+/**
+ * Load balancer config that used to wrap typed ring hash config.
+ */
+class TypedRingHashLbConfig : public Upstream::LoadBalancerConfig {
+public:
+  TypedRingHashLbConfig(
+      const envoy::extensions::load_balancing_policies::ring_hash::v3::RingHash& config);
+
+  const envoy::extensions::load_balancing_policies::ring_hash::v3::RingHash lb_config_;
+};
+
+/**
+ * Implementation of SubsetSelector. This is part of subset load balancer config and is used to
+ * store config of single selector.
+ */
+class SubsetSelectorImpl : public SubsetSelector {
+public:
+  SubsetSelectorImpl(const Protobuf::RepeatedPtrField<std::string>& selector_keys,
+                     envoy::config::cluster::v3::Cluster::LbSubsetConfig::LbSubsetSelector::
+                         LbSubsetSelectorFallbackPolicy fallback_policy,
+                     const Protobuf::RepeatedPtrField<std::string>& fallback_keys_subset,
+                     bool single_host_per_subset);
+
+  // SubsetSelector
+  const std::set<std::string>& selectorKeys() const override { return selector_keys_; }
+  envoy::config::cluster::v3::Cluster::LbSubsetConfig::LbSubsetSelector::
+      LbSubsetSelectorFallbackPolicy
+      fallbackPolicy() const override {
+    return fallback_policy_;
+  }
+  const std::set<std::string>& fallbackKeysSubset() const override { return fallback_keys_subset_; }
+  bool singleHostPerSubset() const override { return single_host_per_subset_; }
+
+private:
+  const std::set<std::string> selector_keys_;
+  const std::set<std::string> fallback_keys_subset_;
+  // Keep small members (bools and enums) at the end of class, to reduce alignment overhead.
+  const envoy::config::cluster::v3::Cluster::LbSubsetConfig::LbSubsetSelector::
+      LbSubsetSelectorFallbackPolicy fallback_policy_;
+  const bool single_host_per_subset_ : 1;
+};
+
+using SubsetLoadbalancingPolicyProto =
+    envoy::extensions::load_balancing_policies::subset::v3::Subset;
+using LegacySubsetLoadbalancingPolicyProto = envoy::config::cluster::v3::Cluster::LbSubsetConfig;
+
+/**
+ * Implementation of LoadBalancerSubsetInfo. Both the legacy and extension subset proto configs
+ * are converted to this class.
+ */
+class LoadBalancerSubsetInfoImpl : public LoadBalancerSubsetInfo {
```

Comment:
ditto here - seems like someone who doesn't include the subset LB would ideally not compile this at all.

Reply:
We can do that after the old path is removed completely.

Six months of waiting. Although I don't  want the waiting actually, but I also have no enough confidence to switch to new path directly.

Reply:
🤔 may be I should be more brave. This PR actually didn't change any lb logic. It just change the way to create lb instance...

I am hesitant now.